README.md
# Grape::Throttling

[![Build Status](https://travis-ci.org/OuYangJinTing/grape-throttling.svg)](https://travis-ci.org/OuYangJinTing/grape-throttling)
[![Gem Version](https://badge.fury.io/rb/grape-throttling.svg)](https://badge.fury.io/rb/grape-throttling)
[![Maintainability](https://api.codeclimate.com/v1/badges/f200c46569aa627315f6/maintainability)](https://codeclimate.com/github/OuYangJinTing/grape-throttling/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/f200c46569aa627315f6/test_coverage)](https://codeclimate.com/github/OuYangJinTing/grape-throttling/test_coverage)

If your api through [`grape`](https://github.com/ruby-grape/grape) build, you can use `grape-throttling` limit api rate.

**Import:** `Grape::Throttling` depend on [`redis`](https://github.com/redis/redis), you must install [`redis`](https://github.com/redis/redis) first

- **Simple**. The you only need to call `use_throttle` in your api internal.
- **Throttle remind**. The response header include `X-RateLimit-Limit`, `X-Ratelimit-Used`, `X-RateLimit-Remaining`, `X-RateLimit-Reset`.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'grape-throttling'
```

And then execute:

```ruby
bundle install
```

Or install it yourself as:

```ruby
gem install grape-throttling
```

## Usage

```ruby
class Api < Grape::API
  use_throttle(max: 60, expire: 1.day, condition: proc { !current_user.is_admin? }, identity: proc { request.ip })
end
```

| Argument Key | Default               | Description                    |
| ------------ | --------------------- | ------------------------------ |
| `max`        | `60`                  | Maximum rate that can be used. |
| `expire`     | `1.day`               | Waiting time for reset rate.   |
| `condition`  | `proc { true }`       | Condition of enabled throttle. |
| `identity`   | `proc { request.ip }` | Used for throttle identity.    |

## Configure

```ruby
# Default configuration
Grape::Throttling.configure do |config|
  config.redis = ::Redis.new(url: 'redis://localhost:6379/0', driver: :hiredis)
  # Api overspeed access message custom method.
  # If doesn't respond，default message is "API rate limit exceeded."
  config.overspeed_message_method = :overspeed_message
end
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at <https://github.com/OuYangJinTing/grape-throttling>. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/OuYangJinTing/grape-throttling/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Grape::Throttling project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/OuYangJinTing/grape-throttling/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Change log

## master

### 0.1.2 (2020-12-24)

- Allow custom overspeed message ([#1](https://github.com/OuYangJinTing/grape-throttling/pull/1/files))

### 0.1.1 (2020-11-05)

- Down grade support ruby >= 2.3.0


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)

require 'pry'
require 'grape/throttling'
require 'minitest/autorun'
require 'minitest/reporters'
require 'rack/test'

Minitest::Reporters.use!


test/grape/api_test.rb
# frozen_string_literal: true

require 'test_helper'

module Grape
  describe API do
    include Rack::Test::Methods

    let(:identity) { :grape }
    let(:max_throttle) { 3 }
    let(:app) do
      max = max_throttle

      Class.new(Grape::API).tap do |grape|
        grape.class_eval do
          singleton_class.attr_reader :throttle
          @throttle = use_throttle(
            max: max,
            condition: proc { params[:condition] },
            identity: proc { params[:identity] }
          )

          helpers do
            def overspeed_message
              'Custom overspeed alert message'
            end
          end

          namespace :namespace do
            post :endpoint do
              status(204)
            end
          end
        end
      end
    end

    after { app.throttle.del(identity) }

    it 'should not counter access time and set throttle headers when condition not satisfied' do
      post '/namespace/endpoint', identity: identity
      assert 204, last_response.status
      assert_equal 0, app.throttle.get(identity)
      assert_nil(last_response.headers.keys.find { |item| item.start_with?('X-RateLimit-') })
    end

    it 'should counter access time and set throttle headers when condition satisfied' do
      post '/namespace/endpoint', identity: identity, condition: true
      assert 204, last_response.status
      assert_equal 1, app.throttle.get(identity)
      assert_equal max_throttle, last_response.headers['X-RateLimit-Limit'].to_i
      assert_equal 1, last_response.headers['X-RateLimit-Used'].to_i
      assert_equal max_throttle - 1, last_response.headers['X-RateLimit-Remaining'].to_i
      assert last_response.headers.key?('X-RateLimit-Reset')
    end

    it 'should limit access and set throttle headers when exceeded max throttle' do
      (max_throttle + 1).times { post '/namespace/endpoint', identity: identity, condition: true }
      assert_equal 403, last_response.status
      assert_equal 'Custom overspeed alert message', last_response.body
      assert_equal max_throttle, last_response.headers['X-RateLimit-Limit'].to_i
      assert_equal max_throttle, last_response.headers['X-RateLimit-Used'].to_i
      assert_equal 0, last_response.headers['X-RateLimit-Remaining'].to_i
      assert last_response.headers.key?('X-RateLimit-Reset')
    end
  end
end


test/grape/request_test.rb
# frozen_string_literal: true

require 'test_helper'

module Grape
  describe Request do
    let(:request) { Request.allocate }

    it '#enabled_throttle' do
      assert_respond_to request, :enabled_throttle
    end

    it '#enabled_throttle=' do
      assert_respond_to request, :enabled_throttle=
    end

    it '#id_throttle' do
      assert_respond_to request, :id_throttle
    end

    it '#id_throttle=' do
      assert_respond_to request, :id_throttle=
    end

    it '#throttle?' do
      assert request.throttle? == true || request.throttle? == false
    end
  end
end


lib/grape/throttling/extension/api.rb
# frozen_string_literal: true

module Grape
  module Throttling
    module Extension
      module Api # :nodoc:
        extend ActiveSupport::Concern

        module ClassMethods # :nodoc:
          # rubocop:disable Style/ClassVars
          def use_throttle(max: 60, expire: 1.day, condition: proc { true }, identity: proc { request.ip }) # rubocop:disable Metrics/MethodLength, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
            max = max.try(:to_i).to_i
            raise ArgumentError, 'max must be positive number' unless max.positive?
            raise ArgumentError, 'condition must be Proc' unless condition.is_a?(Proc)
            raise ArgumentError, 'identity must be Proc' unless identity.is_a?(Proc)

            @@throttle = Counter.new(expire)
            @@throttle.define_singleton_method(:max) { max }
            @@throttle.define_singleton_method(:condition) { condition }
            @@throttle.define_singleton_method(:identity) { identity }

            helpers do
              def mixin_throttle_headers(count)
                header('X-RateLimit-Limit', @@throttle.max.to_s)
                header('X-Ratelimit-Used', count.to_s)
                header('X-RateLimit-Remaining', (@@throttle.max - count).to_s)
                header('X-RateLimit-Reset', (Time.now.utc + @@throttle.ttl(request.id_throttle)).to_s)

                header
              end
            end

            before do
              if instance_eval(&@@throttle.condition)
                request.enabled_throttle = true
                request.id_throttle = instance_eval(&@@throttle.identity)
                count = @@throttle.get(request.id_throttle)

                message = if respond_to?(Throttling.config.overspeed_message_method)
                            send(Throttling.config.overspeed_message_method)
                          else
                            'API rate limit exceeded.'
                          end

                error!(message, 403, mixin_throttle_headers(count)) if count >= @@throttle.max
              end
            end

            after { mixin_throttle_headers(@@throttle.incr(request.id_throttle)) if request.throttle? }

            @@throttle
          end
          # rubocop:enable Style/ClassVars
        end

        Grape::API.include(self)
      end
    end
  end
end
